File: 240322/정수 분배하기/distributing-integers.java
import java.util.*;
import java.io.*;

public class Main {

    static int N,M;
    static int[] nums;

    static boolean isPossible(int k) {
        int cnt=0;
        for(int n: nums) {
            cnt += n/k;
        }
        return cnt>=M?true:false;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        nums = new int[N];
        int r=0;
        for(int i=0;i<N;i++) {
            nums[i] = Integer.parseInt(br.readLine());
            r = Math.max(nums[i],r);
        } //end input

        int k=0,l=0;
        while(l<=r) {
            int mid = (l+r)/2;
            if(mid>0 && isPossible(mid)) {
                k = mid;
                l = mid+1;
            } else {
                r = mid-1;
            }
        }

        System.out.println(k);
    }
}

File: 240913/정수 분배하기/distributing-integers.py
import sys

input = lambda: sys.stdin.readline().rstrip()

#분배가 가능한지 판별하는 함수
def is_possible(m,arr,k):
    cnt=0
    for num in arr:
        cnt += (num//k)
    
    if cnt>=m:
        return True

    return False

def main():
    n,m = map(int, input().split())
    
    arr = []
    r = 0
    for _ in range(n):
        tmp = int(input())
        arr.append(tmp)
        r = max(r,tmp)
    
    #[0~배열내 가장 큰 수] 범위에서 가장 큰 k값 찾기
    k,l = 0,0
    while l<=r:
        mid = (l+r)//2

        if mid>0 and is_possible(m,arr,mid):
            #k=mid일 때 m개를 만드는 분배가 가능한 경우
            k = mid
            l = mid+1
        else:
            r = mid-1
    
    print(k)

    

if __name__ == "__main__":
    main()